src/CSET/operators/read.py

```diff
@@ -17,32 +17,131 @@
+
+    Ensemble data can also be loaded. If it has a realization coordinate
+    already, it will be directly used. If not, it will have its member number
+    guessed from the filename, based on one of several common patterns. For
+    example the pattern *emXX*, where XX is the realization.
+
+    Deterministic data will be loaded with a realization of 0, allowing it to be
+    processed in the same way as ensemble data.
 
     Arguments
     ---------
     loadpath: pathlike
         Path to where .pp/.nc files are located
     constraint: iris.Constraint or iris.ConstraintCombination, optional
-        Constraints to filter by
+        Constraints to filter data by
+    filename_pattern: str, optional
+        Unix shell-style pattern to match filenames to. Defaults to "*"
 
     Returns
     -------
     cubes: iris.cube.CubeList
-        Cubes extracted
+        Cubes loaded
+
+    Raises
+    ------
+    FileNotFoundError
+        If the provided path does not exist
     """
 
+    if isinstance(loadpath, str):
+        loadpath = Path(loadpath)
+
+    if loadpath.is_dir():
+        loadpath = sorted(loadpath.glob(filename_pattern))
+
+    logging.info("Loading file(s): %s", loadpath)
+
     if constraint:
-        return iris.load(loadpath, constraint)
+        cubes = iris.load(loadpath, constraint)
+    else:
+        cubes = iris.load(loadpath)
+    if _is_ensemble(cubes):
+        logging.info("Ensemble data, reloading with correct handling.")
+        if constraint:
+            cubes = iris.load(loadpath, constraint, callback=_ensemble_callback)
+        else:
+            cubes = iris.load(loadpath, callback=_ensemble_callback)
+        cubes.merge()
     else:
-        return iris.load(loadpath)
+        # Give deterministic a realization of 0 so they can be handled in the
+        # same way as ensembles.
+        for cube in cubes:
+            cube.add_aux_coord(
+                iris.coords.AuxCoord(
+                    np.int32(0), standard_name="realization", units="1"
+                )
+            )
+    return cubes
+
+
+def _is_ensemble(cubelist: iris.cube.CubeList) -> bool:
+    """
+    Tests if a cubelist is likely to be ensemble data, or just deterministic
+    data. If cubes either have a realization dimension, or there are multiple
+    files for the same time-step, we can assume it is ensemble data.
+    """
+
+    unique_cubes = set()
+    for cube in cubelist:
+        if cube.coords("realization"):
+            return True
+        # Compare XML representation of cube structure to see if there are
+        # duplicates.
+        cube_content = cube.xml()
+        if cube_content in unique_cubes:
+            logging.debug("Cube Contents: %s", unique_cubes)
+            return True
+        else:
+            unique_cubes.add(cube_content)
+    logging.debug("Cube Contents: %s", unique_cubes)
+    return False
+
+
+def _ensemble_callback(cube, field, filename: str):
```

Comment:
I think this is nice. There are portability concerns as a future TODO, I can see a variety of different filename strings. Perhaps user can specify the filename structure in the suite as a str i.e filenamestructure=mymodelrun_run%Y%M%D_ensemble%E.pp where it can parse this and detect for say %E (i.e. an ensemble)